omero/annotation_scripts/Import_from_csv.py

```diff
@@ -0,0 +1,861 @@
+
+    if total_missing_names > 0:
+        message += f". {total_missing_names} {target_type}(s) not found \
+            (using {'ID' if use_id else 'name'} to identify them)."
+
+    return message, result_obj
+
+
+def get_original_file(omero_obj):
+    """Find last AnnotationFile linked to object if no annotation is given"""
+    file_ann = None
+    for ann in omero_obj.listAnnotations():
+        if ann.OMERO_TYPE == omero.model.FileAnnotationI:
+            file_name = ann.getFile().getName()
+            # Pick file by Ann ID (or name if ID is None)
+            if file_name.endswith(".csv") or file_name.endswith(".tsv"):
+                if (file_ann is None) or (ann.getDate() > file_ann.getDate()):
+                    # Get the most recent file
+                    file_ann = ann
+
+    assert file_ann is not None, \
+        (f"No .csv FileAnnotation was found on {omero_obj.OMERO_CLASS}" +
+         f":{get_obj_name(omero_obj)}:{omero_obj.getId()}")
+
+    return file_ann
+
+
+def read_csv(conn, original_file, delimiter, import_tags):
+    """ Dedicated function to read the CSV file """
+    print("Using FileAnnotation",
+          f"{original_file.id.val}:{original_file.name.val}")
+    provider = DownloadingOriginalFileProvider(conn)
+    # read the csv
+    # Needs omero-py 5.9.1 or later
+
+    try:
+        temp_file = provider.get_original_file_data(original_file)
+        with open(temp_file.name, mode="rt", encoding='utf-8') as f:
+            csv_content = f.readlines()
+    except UnicodeDecodeError as e:
+        assert False, ("Error while reading the csv, convert your " +
+                       "file to utf-8 encoding" +
+                       str(e))
+
+    if delimiter is None:
+        try:
+            # Sniffing on a maximum of four lines
+            delimiter = csv.Sniffer().sniff("\n".join(csv_content[:4]),
+                                            ",;\t").delimiter
+        except Exception as e:
+            assert False, ("Failed to sniff CSV delimiter: " + str(e))
+    rows = list(csv.reader(csv_content, delimiter=delimiter))
+
+    rowlen = len(rows[0])
+    error_msg = (
+        "CSV rows lenght mismatch: Header has {} " +
+        "items, while line {} has {}"
+    )
+    for i in range(1, len(rows)):
+        assert len(rows[i]) == rowlen, error_msg.format(
+            rowlen, i, len(rows[i])
+        )
+
+    # keys are in the header row (first row for no namespaces
+    # second row with namespaces declared)
+    namespaces = []
+    if rows[0][0].lower() == "namespace":
+        namespaces = [el.strip() for el in rows[0]]
+        namespaces = [ns if ns else NSCLIENTMAPANNOTATION for ns in namespaces]
+        rows = rows[1:]
+    header = [el.strip() for el in rows[0]]
+    rows = rows[1:]
+
+    if not import_tags:
+        idx_l = [i for i in range(len(header)) if header[i].lower() != "tag"]
+        header = [header[i] for i in idx_l]
+        namespaces = [namespaces[i] for i in idx_l]
```

Comment:
I got an error here:
```
  File "/opt/omero/server/omero/tmp/omero_omero-server/510/processo47by84e.dir/./script", line 390, in <listcomp>
    namespaces = [namespaces[i] for i in idx_l]
IndexError: list index out of range
```
I chose not to import tags, and I guess that `namespaces` is an empty list at this point since there's no `namespace` in the csv.

Reply:
thank you